[DAG] Fold mul 0 -> 0 when expanding mul into parts.

davemgreen opened the pull request:
If the upper bits are zero, but we expand multiply then immediately convert the multiple into a libcall, there is no opportunity to optimize away the mul. Do so in getNode to make sure extending multiplies optimise cleanly.

s-barannikov commented on llvm/test/CodeGen/RISCV/mul.ll:
```diff
@@ -1829,67 +1829,53 @@ define i64 @mulhsu_i64(i64 %a, i64 %b) nounwind {
-; RV32I-NEXT:    mv a3, s0
+; RV32I-NEXT:    mv a2, s3
+; RV32I-NEXT:    mv a3, s2
 ; RV32I-NEXT:    call __muldi3
```
Note that this multiplies `{a2, a3}` by `s4`, which is either 0 or -1. This is another opportunity for optimization: the multiplication is equivalent to `{-a2 & s4, -a3 & s4}`.
https://alive2.llvm.org/ce/z/C7SJX9

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -11096,10 +11095,14 @@ void TargetLowering::forceExpandMultiply(SelectionDAG &DAG, const SDLoc &dl,
+        ISD::ADD, dl, VT, Hi,
+        DAG.getNode(
+            ISD::ADD, dl, VT,
+            isNullConstant(HiRHS) ? HiRHS
```
It always comes back to yak shaving :/

davemgreen commented:
This should be OK now.

davemgreen commented on llvm/test/CodeGen/AArch64/rem-by-const.ll:
```diff
@@ -279,11 +279,11 @@ define i32 @si32_100(i32 %a, i32 %b) {
-; CHECK-SD-NEXT:    asr x8, x8, #37
-; CHECK-SD-NEXT:    add w8, w8, w8, lsr #31
+; CHECK-SD-NEXT:    asr x9, x8, #37
+; CHECK-SD-NEXT:    add x8, x9, x8, lsr #63
```
It should not be any slower (and could theoretically start the shift before finishing the asr, although I'm not sure that is something that happens in practice).

davemgreen committed:
[DAG] Optimize Mul x, 0 -> 0 in getNode.

This didn't work well on some tests, but they can be fixed with some better
tablegen patterns.

davemgreen commented:
Rebase and ping - thanks.

topperc reviewed (approved):
LGTM
